File: agent_serving/serving/rerank/llm_reranker.py
"""LLMReranker — LLM-based listwise reranking.

Calls ServingLlmClient with a listwise prompt: query + top-N candidates
→ relevance ranking. Falls back gracefully when LLM is unavailable.
"""
from __future__ import annotations

import logging
from typing import Any

from agent_serving.serving.schemas.models import (
    QueryUnderstanding,
    RetrievalCandidate,
)

logger = logging.getLogger(__name__)


class LLMReranker:
    """Listwise LLM reranker with automatic fallback."""

    def __init__(self, llm_client: Any = None) -> None:
        self._llm = llm_client

    async def rerank(
        self,
        candidates: list[RetrievalCandidate],
        understanding: QueryUnderstanding | None = None,
    ) -> list[RetrievalCandidate] | None:
        """Rerank candidates via LLM listwise judgment."""
        if not candidates or not self._llm or not self._llm.is_available():
            return None

        try:
            return await self._try_llm_rerank(candidates, understanding)
        except Exception:
            logger.warning("LLM rerank failed", exc_info=True)
            return None

    async def _try_llm_rerank(
        self,
        candidates: list[RetrievalCandidate],
        understanding: QueryUnderstanding | None,
    ) -> list[RetrievalCandidate]:
        """Execute LLM reranking."""
        top_n = candidates[:20]
        items_text = []
        for i, c in enumerate(top_n):
            text_preview = c.metadata.get("text", "")[:200]
            title = c.metadata.get("title", "")
            items_text.append(f"[{i}] (score={c.score:.3f}) {title}: {text_preview}")

        query = understanding.original_query if understanding else ""

        result = await self._llm.execute(
            pipeline_stage="reranker",
            template_key="serving-reranker",
            input={
                "query": query,
                "candidates": "\n".join(items_text),
                "count": len(top_n),
            },
            expected_output_type="json_object",
        )

        if not result:
            raise ValueError("Empty LLM rerank response")

        inner = result.get("result", result) if isinstance(result, dict) else {}
        parsed = inner.get("parsed_output", {})
        if not parsed:
            raise ValueError("Empty LLM rerank response")

        ranking = parsed.get("ranking", [])
        if not ranking:
            raise ValueError("No ranking in LLM rerank response")

        indexed = {i: c for i, c in enumerate(top_n)}
        reordered: list[RetrievalCandidate] = []

        for item in ranking:
            idx = item.get("index", -1) if isinstance(item, dict) else -1
            if idx in indexed:
                c = indexed.pop(idx)
                new_score = item.get("score", c.score) if isinstance(item, dict) else c.score
                reordered.append(c.model_copy(update={"score": float(new_score)}))

        for c in indexed.values():
            reordered.append(c)

        reordered.extend(candidates[20:])

        return reordered

File: agent_serving_fzl/src/main/java/com/coremasterkb/serving/rerank/LlmReranker.java
package com.coremasterkb.serving.rerank;

import com.coremasterkb.serving.domain.QueryUnderstanding;
import com.coremasterkb.serving.domain.RetrievalCandidate;
import com.coremasterkb.serving.infrastructure.LlmClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

/**
 * Listwise LLM reranker.
 *
 * <p>Sends query + top-N candidate previews to the LLM, which returns a ranking
 * with updated scores. Returns {@code null} on any failure to signal fallback.
 *
 * @see LlmClient#execute(String, String, Map)
 */
public class LlmReranker implements Reranker {

    private static final Logger log = LoggerFactory.getLogger(LlmReranker.class);
    private static final int MAX_CANDIDATES = 20;
    private static final int TEXT_PREVIEW_CHARS = 200;

    private final LlmClient llmClient;

    public LlmReranker(LlmClient llmClient) {
        this.llmClient = llmClient;
    }

    @Override
    public List<RetrievalCandidate> rerank(List<RetrievalCandidate> candidates, QueryUnderstanding understanding) {
        if (candidates == null || candidates.isEmpty()) {
            return null;
        }
        if (llmClient == null || !llmClient.isAvailable()) {
            return null;
        }

        try {
            return doRerank(candidates, understanding);
        } catch (Exception e) {
            log.warn("LLM rerank failed: {}", e.getMessage());
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    private List<RetrievalCandidate> doRerank(List<RetrievalCandidate> candidates, QueryUnderstanding understanding) {
        int topN = Math.min(MAX_CANDIDATES, candidates.size());
        List<RetrievalCandidate> workingSet = candidates.subList(0, topN);

        String itemsText = buildItemsText(workingSet);
        String query = understanding != null ? understanding.originalQuery() : "";

        Map<String, Object> input = new HashMap<>();
        input.put("query", query != null ? query : "");
        input.put("candidates", itemsText);
        input.put("count", topN);

        Map<String, Object> result = llmClient.execute("reranker", "serving-reranker", input);
        if (result == null || result.isEmpty()) {
            throw new IllegalStateException("Empty LLM rerank response");
        }

        // Unwrap envelope: result.result.parsed_output.ranking  or  result.parsed_output.ranking
        Map<String, Object> inner = result;
        Object resultObj = result.get("result");
        if (resultObj instanceof Map<?, ?> m) {
            inner = (Map<String, Object>) m;
        }

        Object parsedObj = inner.get("parsed_output");
        if (!(parsedObj instanceof Map<?, ?> parsed)) {
            throw new IllegalStateException("No parsed_output in LLM rerank response");
        }

        Object rankingObj = parsed.get("ranking");
        if (!(rankingObj instanceof List<?> rankingList) || rankingList.isEmpty()) {
            throw new IllegalStateException("No ranking in LLM rerank response");
        }

        return applyRanking(workingSet, rankingList, candidates, topN);
    }

    @SuppressWarnings("unchecked")
    private List<RetrievalCandidate> applyRanking(
            List<RetrievalCandidate> workingSet,
            List<?> rankingList,
            List<RetrievalCandidate> allCandidates,
            int topN) {

        Map<Integer, RetrievalCandidate> indexed = new HashMap<>();
        for (int i = 0; i < workingSet.size(); i++) {
            indexed.put(i, workingSet.get(i));
        }

        Set<Integer> used = new HashSet<>();
        List<RetrievalCandidate> reordered = new ArrayList<>();

        for (Object item : rankingList) {
            if (!(item instanceof Map<?, ?> m)) continue;
            Map<String, Object> entry = (Map<String, Object>) m;

            Object idxObj = entry.get("index");
            if (!(idxObj instanceof Number)) continue;
            int idx = ((Number) idxObj).intValue();
            if (!indexed.containsKey(idx) || used.contains(idx)) continue;

            used.add(idx);
            RetrievalCandidate c = indexed.get(idx);

            double newScore = c.score();
            Object scoreObj = entry.get("score");
            if (scoreObj instanceof Number n) {
                newScore = n.doubleValue();
            }
            reordered.add(c.withScore(newScore));
        }

        // Append unranked working-set candidates
        for (int i = 0; i < workingSet.size(); i++) {
            if (!used.contains(i)) {
                reordered.add(workingSet.get(i));
            }
        }

        // Append remaining candidates beyond topN
        if (topN < allCandidates.size()) {
            reordered.addAll(allCandidates.subList(topN, allCandidates.size()));
        }

        return reordered;
    }

    private String buildItemsText(List<RetrievalCandidate> candidates) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < candidates.size(); i++) {
            RetrievalCandidate c = candidates.get(i);
            String textPreview = truncate(stringFromMetadata(c, "text"), TEXT_PREVIEW_CHARS);
            String title = stringFromMetadata(c, "title");
            sb.append(String.format("[%d] (score=%.3f) %s: %s%n", i, c.score(), title, textPreview));
        }
        return sb.toString();
    }

    private static String stringFromMetadata(RetrievalCandidate c, String key) {
        Object val = c.metadata().get(key);
        return val instanceof String s ? s : "";
    }

    private static String truncate(String s, int maxLen) {
        if (s == null) return "";
        return s.length() <= maxLen ? s : s.substring(0, maxLen);
    }
}
